JuliaSyntax/src/porcelain/syntax_graph.jl

```diff
@@ -857,3 +875,488 @@ function _map_green_to_ast(parent::Kind, cs::SyntaxList; eq_to_kw=false)
+        # (local (tuple a b c)) => (local a b c)
+        if kind(cs[1]) === K"const"
+            ret_c1_cs = _map_green_to_est(st, preprocessed_green_children(cs[1]))
+            ret_cs = tree_ids(mknode(st, ret_c1_cs))
+            return mknode(cs[1], ret_cs)
+        elseif kind(cs[1]) === K"tuple"
+            cs = preprocessed_green_children(cs[1])
+        end
+    elseif k === K"return" && n_cs === 0
+        push!(cs, coreref("nothing"))
+    elseif k === K"juxtapose"
+        ret_k = K"call"
+        pushfirst!(cs, symleaf("*"))
+    elseif k === K"struct"
+        is_mutable = valleaf(has_flags(st, MUTABLE_FLAG))
+        pushfirst!(cs, is_mutable)
+    elseif k === K"importpath"
+        ret_k = K"."
+        for i in eachindex(cs)
+            if kind(cs[i]) === K"inert"
+                inner_cs = preprocessed_green_children(cs[i])
+                length(inner_cs) === 1 && (cs[i] = only(inner_cs))
+            end
+        end
+    elseif k === K"wrapper" # parse errors only
+        ret_k = K"block"
+    elseif k === K"parameters"
+        kw_in_params = kind(parent) === K"parameters" && parent_i === 1 ?
+            kw_in_params : !(kind(parent) in KSet"vect curly braces ref")
+    elseif k === K"="
+        p_k = kind(parent)
+        because_params = p_k === K"parameters" && parent_i >= 1 && kw_in_params
+        because_call = parent_i > 1 && (p_k == K"ref" ||
+            p_k in KSet"call dotcall" && is_prefix_call(parent))
+        ret_k = because_params || because_call ? K"kw" : K"="
+    end
+
+    # Recurse on `cs`.  If no children change, just return `st`.
+    ret_cs = _map_green_to_est(st, cs; kw_in_params)
+    return ret_cs.ids == children(st).ids && ret_k == kind(st) ?
+        st : setattr!(mknode(st, ret_cs), :kind, ret_k)
+end
+
+function _map_green_to_est(parent::SyntaxTree, cs::SyntaxList;
+                           kw_in_params=false, undef_parent=false)
+    ret_cs = SyntaxList(cs.graph)
+    for (i, c) in enumerate(cs)
+        undef_parent && (i = 0)
+        new_c = _green_to_est(parent, i, c; kw_in_params)
+        @assert should_include_node(new_c)
+        push!(ret_cs, new_c)
+    end
+    ret_cs
+end
+
+# When converting, first delete trivia and wrapper nodes in children so we can
+# observe child kinds before recursing, thus creating fewer "temporary" nodes
+function preprocessed_green_children(st::SyntaxTree)
+    cs = filter(should_include_node, children(st))
+    for i in eachindex(cs)
+        while kind(cs[i]) in KSet"var char parens"
+            inner_cs = preprocessed_green_children(cs[i])
+            if length(inner_cs) === 1
+                cs[i] = inner_cs[1]
+            end
+        end
+    end
+    return cs
+end
+
+# (call f a b (parameters c d) (parameters e)) =>
+# (call f (parameters (parameters e) c d) a b)
+#
+# `mknode` leaks nodes, but having multiple `parameters` blocks is extremely
+# rare nonsense syntax (`f(a,b;c=d;e)`)
```

Comment:
Perhaps move this note into the function, since it's an implementation (not caller) concern